### PR TITLE
Fix weblinks

### LINKED ISSUE #2887: Link provider underline draws 1 extra character when after wide characters
I'm guessing this is in xterm.js as it appears to be purely a rendering issue, the link works file and excludes the `]`:

image

image

### LINKED ISSUE #3539: Web Links Provider shouldn't accept "<>" Characters in opening / ending Clauses
## Details
- Browser and browser version: All
- OS version: All 

### Steps to reproduce

1. Enter `$ echo "\<http://test.com\>"`
2. Hover over the link printed on the next line, which should be `<http://test.com>`
3. Observe that the link that is validated corresponds to `http://test.com>` which doesn't make much sense

Simply adding the `<>` characters to `NEGATED_PATH_CHARACTER_SET` and `NEGATED_DOMAIN_CHARACTER_SET` somewhat resolves this issue, but also has the issue that the test case above won't validate the link inside of the `<>` brackets as well.

### LINKED ISSUE #3721: Wrong weblink highlighting in wrapped terminal lines
Web links in long lines are incorrectly highlighted when the link is on the second line or further.

bugreport

## Details
- Browser and browser version: Chrome
- OS version: MacOS 12.3
- xterm.js version: 4.18.0

### Steps to reproduce

1.  Paste `http://google.com` multiple times into a terminal until it wraps onto a second line
2. Hover over the last instance of the uri.

### LINKED ISSUE #4247: xterm-addon-web-links IViewportRange in hover callback is offset by 1
I have a line with a link:

image

Its the first line (row) of the terminal.

When I select something in the line and call `terminal.getSelectionPosition`, the start and end row is 0, this is what I would expect
When I hover the link and receive the location from the WebLinkProvider hover callback the start and end row is 1

## Details
- Browser and browser version: Electron 21.2.0
- OS version: Windows 11
- xterm.js version: 5.0.0
- xterm-addon-web-links 0.7.0

### Steps to reproduce

1. have a line with text and link
2. have xterm-addon-web-links with hover callback
3. select text, call getSelectionPosition
4. hover link, receive location
5. compare rows of locations

### LINKED ISSUE #4294: Links get wrong position when they are after wide chars with canvas/webgl renderer
See the image below. Note that it only happens to the links detected by the weblinks addon but not OSC 8 links. It also only happens to webgl and canvas renderers but the dom one.
image

I looked at the code a bit. It seems that the web link addon uses the character position (i.e. "哈" is counted 1 even though it is a wide character) as the the value for `IBufferCellPosition.x`. [Code here](https://github.com/xtermjs/xterm.js/blob/497df06f3f848ce1ae2b33dd69b2e56a2791ff3e/addons/xterm-addon-web-links/src/WebLinkProvider.ts#L73). Is this the correct way?

BTW, I am not sure whether this is just #2887.

## Details
- Browser and browser version: Chrome 110
- xterm.js version: Tot

### PR DESCRIPTION
**Pull request by jerch**
This PR tries to tackle several issues regarding weblinks:
- fix underline offsets
- allow more broad url scheme (unicode, with name and password portion)
- use URL() to validate instead of bloated regexp

TODO:
- [x] find better fix for DOM renderer issue
- [x] properly test underline offsets in API tests
- [x] ~fix hover/underline state on wrapped urls not activating~ offtopic here, created #4296 
- [x] better trimming to max 2048 chars

Closes #4294.
Closes #3539.
Closes #4247.
Closes #3721.
Closes #2887.

### CODE REVIEW

File: src/browser/Linkifier2.ts [hunk 1]
```diff
@@ -157,13 +157,15 @@ export class Linkifier2 extends Disposable implements ILinkifier2 {
...
+            // FIXME: commented out due to bug below
+            // this._removeIntersectingLinks(position.y, this._activeProviderReplies);
           }
         });
       }
     }
   }
 
+  // FIXME: What is this supposed to do? Currently it removes wrongly a second link on a wrapped line...
   private _removeIntersectingLinks(y: number, replies: Map<Number, ILinkWithState[] | undefined>): void {
```
Comment (jerch) [inline src/browser/Linkifier2.ts:169]: @Tyriar Could you have a look at this method plz? I think it causes the issue with a second url wrapped to the next not being detected and underlined. I am not quite sure, what the method is supposed to achieve, but with commenting its call out multiple urls wrapping over several lines work is intended.

**Comment by jerch** [comment]
FIXME: Also exclude any sort of brackets from url scheme: <>()[]

File: src/browser/Linkifier2.ts (hunk 1 above)
Comment (jerch) [reply src/browser/Linkifier2.ts:169]: I think this can be tackled separately, thus I created #4296 for it.

**Comment by jerch** [comment]
@Tyriar Imho this is ready for review.

Note that I changed the way how urls are matched - I removed the old explicit positive matching with an exclude matching to allow for newer unicode based urls (supported by all major browsers). Ofc with exclude matching the chance is higher to wrongly match too much, but thats something we can still adjust later on.

The `WebLinkProvider._mapStrIdx` method might be code we want to move to `Buffer` later on (maybe replacing the partially broken `Buffer.stringIndexToBufferIndex`), as we need that functionality in several places.

**Comment by Tyriar** [comment]
vscode-uri is good, but I think it also has some quirks like always encoding the URL when sometimes it's not desirable (eg. encoding `[` and `]`). Could try that out to see how much we get from it? It's an addon so I'm not too concerned about a dependency, especially if it's from our team.

**Comment by Tyriar** [comment]
`new URL` might be useful too?

File: bin/test_weblinks.sh [hunk 2]
```diff
@@ -0,0 +1,34 @@
+#!/bin/bash
+
+# all half width
+echo "aaa http://example.com aaa http://example.com aaa"
```
[the remark this answers is not in the document]
Comment (jerch) [reply bin/test_weblinks.sh:4]: Yes good idea, gonna move it over.

File: addons/xterm-addon-web-links/test/WebLinksAddon.api.ts [hunk 3]
```diff
@@ -64,3 +101,23 @@ async function pollForLinkAtCell(col: number, row: number, value: string): Promi
...
+}
+
+async function resetAndHover(col: number, row: number): Promise<void> {
+  await page.evaluate(`window._customLinkData = [];`);
+  const rowSelector = `.xterm-rows > :nth-child(${row})`;
+  await page.hover(`${rowSelector} > :nth-child(${col})`);
+}
+
+async function evalData(uri: string, range: any): Promise<void> {
+  const data: any[] = await page.evaluate(`window._customLinkData[0]`);
```
[the remark this answers is not in the document]
Comment (jerch) [reply addons/xterm-addon-web-links/test/WebLinksAddon.api.ts:120]: Yes there was some issue with the default type nagging about index access in follow-up code. Can make the type more explicit here.

File: addons/xterm-addon-web-links/src/WebLinksAddon.ts [hunk 4]
```diff
@@ -6,25 +6,19 @@
...
+// NOTE: The repeated end clause is needed to not match a dangling `:`
+// resembling the old (...)*([^:"\'\\s]) final path clause
+// additionally exclude early + final:
+// - unsafe from rfc3986: !*'()
+// - unsafe chars from rfc1738: {}|\^~[]` (minus [] as we need them for ipv6 adresses, also allow ~)
+// also exclude as finals:
+// - final interpunction like ,.!?
+// - any sort of brackets <>()[]{} (not spec conform, but often used to enclose urls)
+// - unsafe chars from rfc1738: {}|\^~[]`
+const strictUrlRegex = /https?:[/]{2}[^\s"'!*(){}|\\\^<>`]*[^\s"':,.!?{}|\\\^~\[\]`()<>]/;
```
[the remark this answers is not in the document]
Comment (jerch) [reply addons/xterm-addon-web-links/src/WebLinksAddon.ts:20]: > I'm guessing people will complain about regressions here

Yes thats the downside of turning the matcher logic upside-down, but I did not find another way while allowing these unicode uris to match. Will see if I get less "overmatching" from the vscode-uri package.

**Comment by jerch** [comment]
@Tyriar Oh well, vscode-uri does not help here much, initially I thought it brings a method to identify url offsets within a string, but it is basically an enhanced `new URL()` nodejs version,  but for that we can just stick to the browser's impl. So nothing to be gained here from vscode-uri.

File: addons/xterm-addon-web-links/src/WebLinksAddon.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (jerch) [reply addons/xterm-addon-web-links/src/WebLinksAddon.ts:20]: Not sure how to go about the risk of overmatching. I think the new matcher is way more capable than the old one, so I'd suggest to go with new one and lets see, if ppl stumble into tons of issues?

And if thats a big issue for peeps, the old addon version would still be around and work as before, so we can fix things iteratively?

**Review by Tyriar — APPROVED** [review]
lgtm, I'll make a note on the release notes about how this is a big change

**Pull request merged**